### DISCUSSION TITLE
Need help building a type-safe custom Prisma wrapper

### DISCUSSION THREAD

**Discussion by adrael** [prisma/prisma#17777] [Q&A]
Hello all!

For a complex project, we're trying to automate [CASL](casl.js.org) (permission management) with Prisma by creating a type-safe wrapper around it.
So far, all our attempts failed:
- client extensions make typescript fail because the Prisma generated types are too complex
- the `$use()` doesn't work for us because it doesn't rely on the session and prevents ut to properly configure the permissions
  - we also tried various options to make `$use()` work such as overrides to pass some context in, or even [node's async hooks](https://nodejs.org/api/async_hooks.html) or [CLS](https://www.npmjs.com/package/node-cls)
- etc.

We are now left with the lesser elegant solution but that works as intended: making a custom Prisma wrapper.
The algorithm is as follows:
- take in a `PrismaClient`
- filter on the keys to only get the models
- create some kind of object that overrides all methods on models (`findUnique`, `findMany`, ...)

This works well, however there is one big caveat - the typings.
We haven't managed so far to make it work in a safe way to make sure the codebase doesn't break.
That is to say that when we make a call to `customPrisma.user.findMany(args);`:
- we have the proper return type just like Prisma does (e.g. `User | null`, or a specific subset of data if `select` is passed in for instance, etc)
- we have the proper checks on the args input just like Prisma does (e.g. you cannot pass in `select` and `include`, you have to pick one, etc)
- and basically everything else prisma does

Can the community help us achieve such a feat?
You'll find below the code we wrote, so far.
Thank you so much for your time and your help 🙏 

P.S: we're also open to any other solution that might fit our needs of course!

```typescript
import { PrismaClient } from "@prisma/client";
import { keyBy, mapValues } from "lodash";

const prisma = new PrismaClient({ ... });

const makeCustomPrisma = (prisma: PrismaClient) => {
  const models = Object.getOwnPropertyNames(prisma).filter((key) => !key.startsWith("_") && !key.startsWith("$"));

  const customModels = mapValues(keyBy(models), (model) => {
    const methods = Object.getOwnPropertyNames(prisma[model]);

    return mapValues(keyBy(methods), (method) => {
      return (args) => {
        // here specific CASL logic will be added
        return prisma[model][method](args);
      };
    });
  });

  return {
    ...prisma,
    ...customModels,
  };
};

const customPrisma = makeCustomPrisma(prisma);
const users = await customPrisma.user.findMany(); // any
```

**Comment by nurul3101** [comment]
To get better context, are you using this `@casl/prisma` [package](https://www.npmjs.com/package/@casl/prisma) to define the permissions on Prisma Models?

**Reply by adrael** [reply]
Absolutely, yes - we're basically trying to build an ability upon session creation, and automatically protecting prisma calls.

We're investigating another way of doing so through Proxy, which allows for transparent typings in the codebase, even if it is not fully type-safe in its definition (which we can accept as it is an isolated piece of code).

```typescript
// This is a bit sad but acceptable as it shouldn't change too much
const PrismaMethods = [
  "findUnique",
  "findUniqueOrThrow",
  "findFirst",
  "findFirstOrThrow",
  "findMany",
  "create",
  "createMany",
  "delete",
  "update",
  "deleteMany",
  "updateMany",
  "upsert",
  "count",
  "aggregate",
  "groupBy",
] as const;

const ability = createPrismaAbility(/* some rules */);
const restrictions = accessibleBy(ability);

new Proxy(prismaClientInstance, {
    get: (target: PrismaClient, property: keyof PrismaClient) => {
      // Ignore non-model properties
      if (property.startsWith("$") || property.startsWith("_")) {
        return target[property];
      }

      return mapValues(target[property], (fn: (args?: unknown | undefined) => unknown, method: PrismaMethod) => {
        return (args?: any) => {
          if (!args) {
            return fn();
          }

          // Simplified on purpose
          args.where = { AND: [args.where, (restrictions as never)[property]] };

          return fn(args);
        };
      });
    },
  });
```
